### ISSUE TITLE
Forbidden parameter name: ?after

### ISSUE THREAD

**Issue by haz** [KCL-Planning/VAL#59]
This is kind of a weird one...it appears as though VAL (and thus the VSCode extension) complains when we use `?after` as a parameter:

![image](https://user-images.githubusercontent.com/225019/219779239-927e71a9-eced-42ec-8353-4aa8a538b842.png)

Any idea what might be going on, @jan-dolejsi ?

**Comment by haz** [comment]
More examples of when it's fine...

![image](https://user-images.githubusercontent.com/225019/219779618-ba2e66dc-b5f1-4b7e-a39f-8537e3881bb0.png)

![image](https://user-images.githubusercontent.com/225019/219779840-c5d46de0-ca2f-42e1-b8d5-5d70284281e8.png)

**Comment by maltehelmert** [comment]
This may be because "after" is a keyword in PDDL preferences syntax. (This shouldn't class with "?after", but perhaps VAL mixes these namespaces.)

**Comment by DerekLong101** [comment]
Malte is right - the tokenizer turns that into a token before the parser gets to see the token stream, so it is already miscategorised as a keyword before it gets to yacc. I can have a look at modifying this behaviour, but in general the keywords will not work as strings for variables, constants, actions, predicates or functions.
________________________________
From: Christian Muise ***@***.***>
Sent: 17 February 2023 19:57
To: KCL-Planning/VAL ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [KCL-Planning/VAL] Forbidden parameter name: ?after (Issue #59)


This is kind of a weird one...it appears as though VAL (and thus the VSCode extension) complains when we use ?after as a parameter:

[image]<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F225019%2F219779239-927e71a9-eced-42ec-8353-4aa8a538b842.png&data=05%7C01%7Cderek.long%40kcl.ac.uk%7C3cede31d71b64877ebfc08db11213ab6%7C8370cf1416f34c16b83c724071654356%7C0%7C0%7C638122606633571437%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=mHPz%2F5CWjfQeUjHkU1by%2BNHPLDg81ROYHjV89hTGxdw%3D&reserved=0>

Any idea what might be going on, @jan-dolejsi<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjan-dolejsi&data=05%7C01%7Cderek.long%40kcl.ac.uk%7C3cede31d71b64877ebfc08db11213ab6%7C8370cf1416f34c16b83c724071654356%7C0%7C0%7C638122606633571437%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=agr6Lz6UMZGUYZHUxxAG6XwtpnRFSfHuj6YKuCK6frQ%3D&reserved=0> ?

**Comment by haz** [comment]
Thanks, both! Even a more descriptive error message would be helpful in this situation (was very much unclear what was wrong with the original action).

What's wild about this one is that it's from an actual IPC domain, and not just a student submission from one of my courses.

**Comment by suddenlyAstral** [comment]
Other forbidden parameters if this trips up anyone:
- The problem name isn't allowed to be "name"
- You may not have a function named (total-time). If that's a predicate, then Validate agrees but Parser marks this as an error

Even some official IPC problems fall for these

**Comment by DerekLong101** [comment]
Yes, VAL's parser is a bit temperamental about name spaces and not very good at reporting the violations it perceives. One of these days I might get back to spending some time looking at it... One of those possible "retirement projects" that seem so plausible as retirement approaches and probably a lot less appealing afterwards...
________________________________
